core/lib/zksync_core/src/utils/mod.rs

```diff
@@ -126,6 +126,31 @@ pub(crate) async fn projected_first_l1_batch(
     Ok(snapshot_recovery.map_or(L1BatchNumber(0), |recovery| recovery.l1_batch_number + 1))
 }
 
+/// Obtains a protocol version projected to be applied for the next miniblock. This is either the version used by the last
+/// sealed miniblock, or (if there are no miniblocks), one referenced in the snapshot recovery record.
+pub(crate) async fn pending_protocol_version(
```

Comment:
There are two potential problems: 
What if there are no miniblocks yet or on goerli for very ancient blocks there is no protocol version set.

Reply:
- The case with no miniblocks after snapshot recovery *is* being handled, so this leaves the case without blocks and without snapshot recovery. I don't think we can handle such a case; realistically, it would be most probably caused by `pending_protocol_version()` called "too early" in the node lifecycle. @RomanBrodetski WDYT?
- We load the *latest* miniblock header, so no protocol version for some *ancient* miniblocks is not a concern. The latest miniblock on all nodes should have its protocol version set, right? Even if they don't, it's not a big deal: we could use `ProtocolVersionId::last_potentially_undefined()` as a fallback. This would make `pending_protocol_version()` completely equivalent to the current implementation.

Reply:
> we could use ProtocolVersionId::last_potentially_undefined() as a fallback

It w'd be great

Reply:
> so this leaves the case without blocks and without snapshot recovery

So it can only happen if we run state keeper before genesis (or snapshotshot recovery) was complete. I think we need to be more explicit in expectations that component can have, but currently I think the state keeper behaviour is undefined. Now, I can imagine a problem if this method is called from e.g. snapshot recovery logic or genesis logic (as @slowli mentioned, too early in node lifecycle). This shouldn't happen, but it still concerns me that we have methods with undefined behaviour for some DB states.

I guess the easiest would be to call this pre-genesis state (point `1` [from this slack message](https://matter-labs-workspace.slack.com/archives/C05ANUCGCKV/p1707226824450369?thread_ts=1707209111.843459&cid=C05ANUCGCKV)) a special case that the components don't have to handle. This way they can always rely on the fact that there are `miniblocks`/`l1_batches` OR `snapshot_reovery` entires.

Reply:
> undefined behaviour

Nitpick: this isn't UB in Rust terms, and I think returning an error with intelligible cause is the best possible course of action in such a case, which this logic does.